### DISCUSSION TITLE
(Done) August 13th Meetup: Superoptimization

### DISCUSSION THREAD

**Discussion by frenchy64** [madclj/madclj.com#17] [Announcements]
👍 to RSVP

https://madclj.com/

Wednesday, August 13th, 2025
6:30 PM to 9:00 PM CT

[StartingBlock Madison](https://www.google.com/maps/search/?api=1&query=43.081245%2C%20-89.374054)
821 E Washington Ave 2nd floor · Madison, WI

Let's discuss [superoptimization](https://github.com/orgs/madclj/discussions/17#discussioncomment-14064503) and its [Clojure implementation](https://github.com/twhume/superoptimiser/tree/master/SuperOptimiser).

**Comment by sivakusayan** [comment]
I thought this was interesting and just wanted to share. 🙂

There is apparently a concept known as [superoptimization](https://en.wikipedia.org/wiki/Superoptimization), where a compiler will try to find the most "optimal" generated code for a given input program, potentially at the cost of compilation speed. [The paper that coined the term](https://web.stanford.edu/class/cs343/resources/superoptimizer.pdf) seems to define "optimal" as minimizing code size, although I didn't check if more recent, related projects have the same definition of "optimal".

It seems that someone wrote [a superoptimizer for the JVM](https://onlinelibrary.wiley.com/doi/10.1002/spe.2240) in Clojure! They also seem to define "optimal" as minimizing code size. [Here is the code](https://github.com/twhume/superoptimiser/tree/master/SuperOptimiser). I haven't played around with it, and I'm guessing it's too slow to use in a JIT. But perhaps it could inspire better JVM optimization passes by comparing its output with the output of traditional passes.

**Reply by frenchy64** [reply]
That's cool! I'll use that as the discussion for the meetup.

**Reply by frenchy64** [reply]
I updated the code to Clojure 1.12, but I doubt it still works since the ASM version is different. Maybe it can be upgraded https://github.com/frenchy64/superoptimiser/pull/1/files

The [original paper's](https://web.stanford.edu/class/cs343/resources/superoptimizer.pdf) signum example was ported to Clojure [here](https://github.com/frenchy64/superoptimiser/blob/5b31b5ab015fbfc703bba29ecf0f724cfd5cc664/SuperOptimiser/src/Drivers/Signum.clj).

**Reply by frenchy64** [reply]
Here's what we coded up during the meetup https://github.com/frenchy64/superoptimiser/pull/2/files
